mygrad/nnet/layers/recurrent.py

```diff
@@ -158,3 +159,158 @@ def simple_RNN(X, U, W, s0=None, bp_lim=None):
+        partial dL / ds(t+1) * ds(t+1) / dh(t) * dh(t) / dr(t) * dr(t) / ds(t)
+    """
+    dz = 1 - z # note: not actually ds(t+1) / dz(t)
+    dh = 1 - h ** 2
+    dLdh = np.dot(dLds * dh * dz, Wh.T)
+
+    return z * dLds + \
+           np.dot((dLds * (s - h)) * z * dz, Wz.T) + \
+           dLdh * r + \
+           np.dot(dLdh * s * r * (1 - r), Wr.T)
+
+
+class GRUnit(Operation):
+    def __init__(self, Uz, Wz, bz, Ur, Wr, br, Uh, Wh, bh, bp_lim):
+        self.Uz = Uz
+        self.Wz = Wz
+        self.bz = bz
+
+        self.Ur = Ur
+        self.Wr = Wr
+        self.br = br
+
+        self.Uh = Uh
+        self.Wh = Wh
+        self.bh = bh
+
+        self.bp_lim = bp_lim
+
+        self._input_seq = None
+        self._hidden_seq = []
+
+        self._z = []
+        self._r = []
+        self._h = []
+
+
+    def __call__(self, seq, s0=None):
+        self._input_seq = seq if self._input_seq is None else np.vstack((self._input_seq, seq))
+
+        z = np.zeros((seq.shape[0], seq.shape[1], self.Uz.shape[-1]))
+        r = np.zeros((seq.shape[0], seq.shape[1], self.Ur.shape[-1]))
+        h = np.zeros((seq.shape[0], seq.shape[1], self.Uh.shape[-1]))
+        out = np.zeros((seq.shape[0] + 1, seq.shape[1], self.Uz.shape[-1]))
+
+        if self._hidden_seq:
+            out[0] = self._hidden_seq[-1].data
+        elif s0 is not None:
+            out[0] = s0.data if isinstance(s0, Tensor) else s0
+
+        np.dot(seq, self.Uz.data, out=z)
+        np.dot(seq, self.Ur.data, out=r)
+        np.dot(seq, self.Uh.data, out=h)
+
+        _gru_layer(out, z, r, h,
+                   self.Wz.data, self.Wr.data, self.Wh.data,
+                   self.bz.data, self.br.data, self.bh.data)
+
+
+        if not self._hidden_seq:
+            self._hidden_seq = Tensor(out, _creator=self)
+        else:
+            new_dat = np.vstack((self._hidden_seq.data, out[1:]))
+            self._hidden_seq = Tensor(new_dat, _creator=self)
+
+        if not self._z:
+            self._z = Tensor(z, _creator=self)
+        else:
+            new_dat = np.vstack((self._z.data, z[1:]))
+            self._z = Tensor(new_dat, _creator=self)
+
+        if not self._r:
+            self._r = Tensor(r, _creator=self)
+        else:
+            new_dat = np.vstack((self._r.data, r[1:]))
+            self._r = Tensor(new_dat, _creator=self)
+
+        if not self._h:
+            self._h = Tensor(h, _creator=self)
+        else:
+            new_dat = np.vstack((self._h.data, h[1:]))
+            self._h = Tensor(new_dat, _creator=self)
+
+        return self._hidden_seq
+
+
+    def backward(self, grad, seq_index=None):
+        s = self._hidden_seq.data[:-1]
+        z = self._z.data
+        r = self._r.data
+        h = self._h.data
+
+        dLds = grad[1:]
+
+        for i in range(min(s.shape[0] - 1, self.bp_lim)):
+            #  dL(t) / ds(t) + dL(t+1) / ds(t)
+            dLds[len(dLds) - (i + 2)] += _gru_dLds(s[len(dLds) - (i + 1)],
```

Comment:
I went through the math again and agree. I was having problems with the simple-rnn slice approach earlier, but just implemented a fix and it is now working properly (slicing was an index off for the s, z, r, and h parameters to _gru_dLds)